### ISSUE TITLE
Time management bug?

### ISSUE THREAD

**Issue by Marcin1960** [gcp/leela-zero#1185]
I used big net, without --timemanage option. My hardware is weak, so trying to give more time is critical.

At my surprise my bot having 20 second for a move was leaving early almost each move, after 10 second or 7 seconds and losing game after game. Guys told me that the best move was already found by timemanage so nothing can be done about.

OK, I put "--timemanage off". Sure enough all 20 sec were used and my bot started to win!

The conclusion - timemanage evaluation (when to leave early), does not work!

**Comment by Marcin1960** [comment]
BTW, I get this strange discrepancy in evaluation output:

Thinking at most 18.5 seconds...
NN eval=0.335862
Playouts: 251, Win: 66.52%, PV: O17 N16 P19 S15 S14 S16 ....
Playouts: 488, Win: 66.61%, PV: O17 N16 P19 S15 S14 S16 ...

When "Win:" gets into 90s%, NN is in single digits. Is it correct?

**Comment by Marcin1960** [comment]
It happened each time, out of 10 games maybe.

Not always big evidence is necessary, sometimes logical reasoning should suffice. If I use big net like 192x15 on my PC I can get 2000 playouts in 20 seconds with "--timemanage off".

Otherwise, I get about 1000 playouts most of the moves. Is it possible that so few playouts are sufficient?
When I gave 2 minutes per move with 20x256 net, the engine was leaving early most of the moves after about 60 seconds.

**Comment by gcp** [comment]
The point of timemanage is:

1) To stop thinking when it's 100% sure the best move can't change.
2) To leave more time for the next moves.

So "logical reasoning" says this must be invalid. I might be wrong, but you need data, not vague anecdotes.

**Comment by Marcin1960** [comment]
What type of data do you need?

**Comment by killerducky** [comment]
Marcin1960, I recommend you just run with `--timemanage off`. I believe you are running with very little main time, and byoyomi period of 20 seconds. In this mode the advantage of using (almost) all 20 seconds is to push more moves into the tree for reuse in the next move. The disadvantage is your opponent can think more during that time also, and your human opponent might be annoying you spend extra timing thinking on "obvious" replies. Note that many strong humans use their entire 20 seconds even if they decide on their move very early on.

Also note that with large main times it's more complicated what you should do. I think I saw another issue discussing changing how much time is allocated per move in that case.

An enhancement would be a 3rd mode where moves are pruned, as in timemanage. When down to only 1 choice, continue thinking only about that move until time is up, so that more visits are pushed into the tree for reuse.

**Comment by gcp** [comment]
>I believe you are running with very little main time, and byoyomi period of 20 seconds.

Oh, so there's indeed no point in using time management then. Given the large prediction rate we have right now, I think we should disable it (by default) in byo yomi time?

**Comment by killerducky** [comment]
Yes, this raises the annoying question of how much configuration to allow. Configurable timemanage on/off during main, and on/off during byoyomi?

Personally when playing or observing it play in 20s byoyomis I prefer it to move quicker in obvious situations and I don't mind a small decrease in strength

**Comment by gcp** [comment]
I think auto should turn it off during byo-yomi, but having it on or off forces it enabled or disabled.

**Comment by gcp** [comment]
Put another way, you could wonder what cases you'd ever have where you'd want timemanage off. I can see the point for playouts/visits testing, but otherwise?

**Comment by gcp** [comment]
>Personally when playing or observing it play in 20s byoyomis I prefer it to move quicker in obvious situations and I don't mind a small decrease in strength

I know what you mean but personally I consider it a huge problem with byo yomi time. At least from the computer perspective, there's never any other optimal method aside from using all your time.

**Comment by Marcin1960** [comment]
I am using --timemanage off right now. It gets about 1 or 2 stones stronger. I am all set :)

Just wanted to turn you attention to. I suspect that evaluation might be broken if it comes to the false conclusions that the best move is already found, and much better moves still can come.

**Comment by gcp** [comment]
>I suspect that evaluation might be broken if it comes to the false conclusions that the best move is already found, and much better moves still can come.

There's no evidence this is the case.

The problem is that time management works on the assumption that moving fast is good because it saves time. That's not true in (Japanese) byo yomi.

**Comment by Marcin1960** [comment]
It would be nice if better move were picked with  time management on.

I wonder how the decision is being made, and if it used for other things than time management.

**Comment by tterava** [comment]
The evaluation works as it should and if LZ plays a move without using all of its time, it would have played the same move even with --timemanage off and having spent all of its time.

Spending more time (even when LZ already knows the move it will play) allows it to find better moves on subsequent turns however. That way it's more prepared for opponent response and might find moves it wouldn't if it had moved instantly previous turn.

I think the only change that should be made for best performance is to have time management turn off when byo yomi is reached. It seems GTP doesn't support multiple byo yomi periods so there's no need to worry about those.

**Comment by tterava** [comment]
I actually just realized that the strength could be improved even if time management is turned off:
Here's how:
As with time management on, calculate the n/s at the start of a turn. Keep checking if any move has reached so many playouts that other moves cannot surpass it. Instead of making that move instantly, spend the rest of the time focused only on that move.

I think currently LZ wastes playouts on moves it knows it is not going to play.

**Comment by tterava** [comment]
@Marcin1960 
Just to clarify: On an unexplored move it makes no difference if timemanage is turned on or off. On byo yomi it does make a difference on the following moves though and that's why timemanagement off is stronger.

**Comment by killerducky** [comment]
So we can change the definition of `--timemanage auto` to be: Do whatever is best to attain maximum strength. This means keeping the same behavior during maintime (there are some optimizations to be had here too though), but use all time in byoyomi. Pruning is still done, when down to one move put all playouts into it for building the tree.

`--timemanage on` can remain the same behavior as before for people who just want the computer to move a little faster in Japanese byo yomi despite the loss of strength.

`--timemanage off` is mostly for people who want to see 2nd, 3rd, etc best moves analyzed, for example using LZ to analyze games.

**Comment by TFiFiE** [comment]
The problem is more general than this. The misleadingly named `TimeControl::max_time_for_move` does time management of its own and then returns a single value, which also makes it impossible to spend more rather than less time than usual on a move (e.g. because the most visited child is currently not also the best scoring one). Rather than doing any time management of its own, `TimeControl` should simply supply objective info to some (centralized) code actually responsible for the time management, which would also have dynamic search data to determine when to make a move.

**Comment by alreadydone** [comment]
So some [old idea](https://github.com/gcp/leela-zero/issues/951#comment-369717994) which I thought was implemented by @TFiFiE in https://github.com/gcp/leela-zero/pull/968 wasn't actually implemented and it seems there are some difficulties...

**Comment by gcp** [comment]
>Rather than doing any time management of its own, TimeControl should simply supply objective info to some (centralized) code actually responsible for the time management, which would also have dynamic search data to determine when to make a move.

You can move the logic somewhere else and have it return a tuple of the things that interest you, which is going to be some combination of max_time, target_time, special flags like byo yomi, etc, or have the search call in more specific methods of TimeControl to calculate what is appropriate. 

The point I'm making is that the "objective info" that you want to supply is likely going to be most of what TimeControl knows, so it might as well have the logic.

**Comment by tterava** [comment]
If roy7's new move selection strategy is implemented then it will have big implications on time control as well. You can't know in advance which move will get selected based on visits so all optimizations which rely on that assumption will be invalid.

**Comment by Marcin1960** [comment]
I would like one thing, that the time management were doing one think. Instead of leaving in the half of allotted byoyomi time (when I gave 2 minutes, it was usually leaving after about 60 seconds, in case of 20 seconds leaving after 10 seconds), was  leaving after finding a really good move, no matter if in half time or 80% or 20%).

**Comment by gcp** [comment]
>I would like one thing, that the time management were doing this:...was leaving after finding a really good move, no matter if in half time or 80% or 20%).

That's what it's doing, it's just that in practice it can't be sure until it's halfway. And as already explained, in byo yomi times this makes the program weaker. It's optimal to use all time, always.

**Comment by Marcin1960** [comment]
"in practice it can't be sure until it's halfway"

Is it really sure?  Is it possible that there is some undetected bug in evaluation code? Or does it mean that it just picks the best move found in half time? That how is it different from the  simply time limit reduced  by half??

It would be nice if time management worked for byo yomi as well.

**Comment by roy7** [comment]
@Marcin1960 It's not the same as using half the time all the time because it depends on the visit count of the most visited node. If many nodes are visited, much of the visit count limit might be used. But if all visits are going into a single primary variation, then once it's used have the visit count it will stop. 

Even without timemanage, --visits will still cap the max visits before it will make a move, time remaining or not. If you want max strength on a bot playing competitively, I imagine you'd not use --visits or --playouts, and turn on pondering as well.

If you don't want to ponder but still want some sort of cap so the bot doesn't think too long in main time, --playouts could do that. Then it'll add that many playouts every time it moves instead of stopping early if it hits a visit count.
